test/extensions/filters/network/local_ratelimit/local_ratelimit_test.cc

```diff
@@ -0,0 +1,199 @@
+#include "envoy/config/filter/network/local_rate_limit/v2alpha/local_rate_limit.pb.validate.h"
+
+#include "common/stats/isolated_store_impl.h"
+
+#include "extensions/filters/network/local_ratelimit/local_ratelimit.h"
+
+#include "test/mocks/event/mocks.h"
+#include "test/mocks/network/mocks.h"
+#include "test/mocks/runtime/mocks.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::_;
+using testing::InSequence;
+using testing::NiceMock;
+using testing::Return;
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace LocalRateLimitFilter {
+namespace {
+
+class LocalRateLimitTestBase : public testing::Test {
+public:
+  void setup(const std::string& filter_yaml) {
+    envoy::config::filter::network::local_rate_limit::v2alpha::LocalRateLimit proto_config;
+    TestUtility::loadFromYamlAndValidate(filter_yaml, proto_config);
+    EXPECT_CALL(*fill_timer_, enableTimer(_, nullptr));
+    config_ = std::make_shared<Config>(proto_config, dispatcher_, stats_store_, runtime_);
+  }
+
+  NiceMock<Event::MockDispatcher> dispatcher_;
+  Stats::IsolatedStoreImpl stats_store_;
+  NiceMock<Runtime::MockLoader> runtime_;
+  Event::MockTimer* fill_timer_ = new Event::MockTimer(&dispatcher_);
+  ConfigSharedPtr config_;
+};
+
+// Verify token bucket functionality.
+TEST_F(LocalRateLimitTestBase, TokenBucket) {
```

Comment:
Given the filter itself implements the token bucket's threading model, could we have testing done in here were we exercise the multi-threaded nature of the code. Have more then one thread trying to get a connection and only one being able to do so?

Reply:
See my other comment. I will add deterministic tests.